lib/sdf-server/src/garbage_collection.rs

```diff
@@ -0,0 +1,156 @@
+        task_tracker: &TaskTracker,
+        task_token: CancellationToken,
+    ) -> Result<Self> {
+        let (services_context, layer_db_graceful_shutdown) =
+            init::services_context_from_config(&config, task_token).await?;
+
+        task_tracker.spawn(layer_db_graceful_shutdown.into_future());
+
+        Ok(Self { services_context })
+    }
+
+    #[instrument(
+        name = "sdf.snapshot_garbage_collector.garbage_collect_snapshots",
+        level = "info",
+        skip_all
+    )]
+    pub async fn garbage_collect_snapshots(self) -> Result<()> {
+        let span = current_span_for_instrument_at!("info");
+
+        let dal_context = self.services_context.clone().into_builder(true);
+        let ctx = dal_context
+            .build_default(None)
+            .await
+            .map_err(|err| span.record_err(err))?;
+        let ctx = &ctx;
+
+        let mut open_change_set_snapshot_ids = HashSet::new();
+        let mut all_snapshot_ids = HashSet::new();
+
+        // Gather the WorkspaceSnapshotAddress of all open change sets.
+        let open_statuses: Vec<String> = ChangeSetStatus::iter()
+            .filter_map(|status| {
+                if status.is_active() {
+                    Some(status.to_string())
+                } else {
+                    None
+                }
+            })
+            .collect();
+        info!(
+            "Change set status(es) to consider 'active': {:?}",
+            &open_statuses
+        );
+        let change_set_snapshot_rows = ctx.txns().await?.pg().query(
+            "SELECT workspace_snapshot_address AS snapshot_id FROM change_set_pointers WHERE status = ANY($1::text[]) GROUP BY workspace_snapshot_address",
+            &[&open_statuses],
+        ).await?;
+        for row in change_set_snapshot_rows {
+            let snapshot_id: WorkspaceSnapshotAddress = row.try_get("snapshot_id")?;
+            open_change_set_snapshot_ids.insert(snapshot_id);
+        }
+        info!(
+            "Found {} distinct snapshot address(es) for open change sets.",
+            open_change_set_snapshot_ids.len()
+        );
+
+        // Gather the WorkspaceSnapshotAddress of all existing snapshots that are
+        // at least an hour old.
+        //
+        // By only considering the snapshots that are at least an hour old, we avoid
+        // race conditions where a change set is created, or modified between when we
+        // queried the change_set_pointers table and the workspace_snapshots table.
+        // We can't rely on transactional integrity to avoid race conditions as the
+        // tables are in completely separate databases.
+        let snapshot_id_rows = ctx
+            .layer_db()
+            .workspace_snapshot()
+            .cache
+            .pg()
+            .query(
+                "SELECT key AS snapshot_id FROM workspace_snapshots WHERE created_at < NOW() - '1 hour'::interval GROUP BY key",
+                &[],
+            )
```

Comment:
Confirmed that this is running a SQL query on a pg client (out of the pool) in a non-transaction query. No other side effects!